src/Form/AuthenticationForm.php

```diff
@@ -372,26 +392,198 @@ public function validateForm(array &$form, FormStateInterface $form_state) {
-        $form_state->setError($form, $this->t('Connection failed. Response from Apigee Edge: %response<br>For further information please check the <a href=":url">log messages</a>.', [
-          '%response' => $exception->getMessage(),
-          ':url' => Url::fromRoute('dblog.overview')->toString(),
-        ]));
+
+      $form_state->setError($form, $this->t('@suggestion Error message: %response.', [
+        '@suggestion' => $this->createSuggestion($exception, $key),
+        '%response' => $exception->getMessage(),
+      ]));
+
+      // Display debug information.
+      $form['debug']['#access'] = $form['debug']['debug_text']['#access'] = TRUE;
+      $form['debug']['debug_text']['#value'] = $this->createDebugText($exception, $key, $key_token);
+    }
+  }
+
+  /**
+   * Creates a suggestion text to be displayed in the connection failed message.
+   *
+   * @param \Exception $exception
+   *   The thrown exception during form validation.
+   * @param \Drupal\key\KeyInterface $key
+   *   The used key during form validation.
+   *
+   * @return \Drupal\Component\Render\MarkupInterface
+   *   The suggestion text to be displayed.
+   */
+  protected function createSuggestion(\Exception $exception, KeyInterface $key): MarkupInterface {
+    /** @var \Drupal\apigee_edge\Plugin\KeyType\BasicAuthKeyType $key_type */
+    $key_type = $key->getKeyType();
+
+    // Failed to connect to the Oauth authorization server.
+    if ($exception instanceof OauthAuthenticationException) {
+      /** @var \Drupal\apigee_edge\Plugin\KeyType\OauthKeyType $key_type */
+      $fail_text = $this->t('Failed to connect to the OAuth authorization server.');
+      // General error message.
+      $suggestion = $this->t('@fail_text Check the debug information below for more details.', [
+        '@fail_text' => $fail_text,
+      ]);
+      // Invalid credentials.
+      if ($exception->getCode() === 401) {
+        // Invalid credentials using defined client_id/client_secret.
+        if ($key_type->getClientId($key) !== Oauth::DEFAULT_CLIENT_ID || $key_type->getClientSecret($key) !== Oauth::DEFAULT_CLIENT_SECRET) {
+          $suggestion = $this->t('@fail_text The given username (%username) or password or client ID (%client_id) or client secret is incorrect.', [
+            '@fail_text' => $fail_text,
+            '%client_id' => $key_type->getClientId($key) ?? Oauth::DEFAULT_CLIENT_ID,
+            '%username' => $key_type->getUsername($key),
+          ]);
+        }
+        // Invalid credentials using default client_id/client_secret.
+        else {
+          $suggestion = $this->t('@fail_text The given username (%username) or password is incorrect.', [
+            '@fail_text' => $fail_text,
+            '%username' => $key_type->getUsername($key),
+          ]);
+        }
+      }
+      // Failed request.
+      elseif ($exception->getCode() === 0) {
+        if ($exception->getMessage() === 'Host can not be empty.') {
```

Comment:
Message should be defined in a constant.

Reply:
Btw, isn't there anything else specific in the exception that you can use in the condition? I do not really prefer writing a condition on an message that could change easily.